unstable test checkpoint_race_ddl_crash

### What did you do?

https://prow.tidb.net/jenkins/job/pingcap/job/ticdc/job/pull_cdc_mysql_integration_heavy_next_gen/244/stages/?selected-node=417

### What did you expect to see?

_No response_

### What did you see instead?

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x4d35381]
goroutine 728 [running]:
github.com/pingcap/ticdc/pkg/common.(*TableInfo).GetColumns(...)
	github.com/pingcap/ticdc/pkg/common/table_info.go:206
github.com/pingcap/ticdc/pkg/sink/mysql.ddlSessionTimestampFromOriginDefault(0xc000d02300, {0x0, 0x0})
	github.com/pingcap/ticdc/pkg/sink/mysql/helper.go:638 +0x81
github.com/pingcap/ticdc/pkg/sink/mysql.(*Writer).execDDL(0xc0006f5900, 0xc000d02300)
	github.com/pingcap/ticdc/pkg/sink/mysql/mysql_writer_ddl.go:108 +0xca8
github.com/pingcap/ticdc/pkg/sink/mysql.(*Writer).execDDLWithMaxRetries.func1.1()
	github.com/pingcap/ticdc/pkg/sink/mysql/mysql_writer_ddl.go:197 +0x77
github.com/pingcap/ticdc/pkg/metrics.(*Statistics).RecordDDLExecution(0xc0015045a0, 0xc0018b9768)
	github.com/pingcap/ticdc/pkg/metrics/statistics.go:104 +0xc6
github.com/pingcap/ticdc/pkg/sink/mysql.(*Writer).execDDLWithMaxRetries.func1()
	github.com/pingcap/ticdc/pkg/sink/mysql/mysql_writer_ddl.go:195 +0x7d
github.com/pingcap/ticdc/pkg/retry.run({0x7389490, 0xc000a85810}, 0xc004973aa0, 0xc0050ea600)
	github.com/pingcap/ticdc/pkg/retry/retry_with_opt.go:57 +0xf7
github.com/pingcap/ticdc/pkg/retry.Do({0x7389490, 0xc000a85810}, 0xc0018b9aa0, {0xc0018b9a80?, 0x667cfe5?, 0x9?})
	github.com/pingcap/ticdc/pkg/retry/retry_with_opt.go:34 +0x47
github.com/pingcap/ticdc/pkg/sink/mysql.(*Writer).execDDLWithMaxRetries(0xc0006f5900, 0xc000d02300)
	github.com/pingcap/ticdc/pkg/sink/mysql/mysql_writer_ddl.go:194 +0x134
github.com/pingcap/ticdc/pkg/sink/mysql.(*Writer).FlushDDLEvent(0xc0006f5900, 0xc000d02300)
	github.com/pingcap/ticdc/pkg/sink/mysql/mysql_writer.go:150 +0x74
github.com/pingcap/ticdc/downstreamadapter/sink/mysql.(*Sink).WriteBlockEvent(0xc001508f00, {0x73e3da0, 0xc000d02300})
	github.com/pingcap/ticdc/downstreamadapter/sink/mysql/sink.go:299 +0x129
github.com/pingcap/ticdc/pkg/applier.(*RedoApplier).applyDDL(0xc000a6c870, {0x7389490, 0xc000a6c8c0}, 0xc001ec4d20, 0x0?)
	github.com/pingcap/ticdc/pkg/applier/redo.go:331 +0x2c6
github.com/pingcap/ticdc/pkg/applier.(*RedoApplier).consumeLogs(0xc000a6c870, {0x7389490, 0xc000a6c8c0})
	github.com/pingcap/ticdc/pkg/applier/redo.go:192 +0x366
github.com/pingcap/ticdc/pkg/applier.(*RedoApplier).Apply.func3()
	github.com/pingcap/ticdc/pkg/applier/redo.go:482 +0x59
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync@v0.17.0/errgroup/errgroup.go:93 +0x50
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 1
	golang.org/x/sync@v0.17.0/errgroup/errgroup.go:78 +0x95
script returned exit code 2
```

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
(paste TiDB cluster version here)
```

Upstream TiKV version (execute `tikv-server --version`):

```console
(paste TiKV version here)
```

TiCDC version (execute `cdc version`):

```console
(paste TiCDC version here)
```

--- a/pkg/common/event/redo.go
+++ b/pkg/common/event/redo.go
@@ -403,7 +403,7 @@ func (r *RedoDDLEvent) ToDDLEvent() *DDLEvent {
 		}
 		columns = append(columns, colInfo)
 	}
-	tableInfo := commonType.NewTableInfo4Decoder(r.TableName.Schema, &timodel.TableInfo{
+	tableInfo := commonType.WrapTableInfo(r.TableName.Schema, &timodel.TableInfo{
 		ID:      r.TableName.TableID,
 		Name:    ast.NewCIStr(r.TableName.Table),
 		Columns: columns,

--- a/pkg/common/event/redo.go
+++ b/pkg/common/event/redo.go
@@ -77,10 +77,23 @@ type DDLEventInRedoLog struct {
 	StartTs           uint64            `msg:"start-ts"`
 	CommitTs          uint64            `msg:"commit-ts"`
 	Query             string            `msg:"query"`
+	Columns           []*ColumnInfo     `msg:"columns"`
 	BlockedTables     *InfluencedTables `msg:"blocked-tables"`
 	BlockedTableNames []SchemaTableName `msg:"blocked-table-names"`
 	NeedDroppedTables *InfluencedTables `msg:"need-dropped-tables"`
 	NeedAddedTables   []Table           `msg:"need_added_tables"`
+}
+
+type ColumnInfo struct {
+	Name               string `msg:"name"`
+	OriginDefaultValue any    `msg:"origin_default"`
+	Type               byte   `msg:"type"`
+	// Version means the version of the column info.
+	// Version = 0: For OriginDefaultValue and DefaultValue of timestamp column will stores the default time in system time zone.
+	//              That is a bug if multiple TiDB servers in different system time zone.
+	// Version = 1: For OriginDefaultValue and DefaultValue of timestamp column will stores the default time in UTC time zone.
+	//              This will fix bug in version 0. For compatibility with version 0, we add version field in column info struct.
+	Version uint64 `msg:"version"`
 }
 
 // RedoColumn is for column meta
@@ -203,12 +216,22 @@ func (r *RedoRowEvent) ToRedoLog() *RedoLog {
 
 // ToRedoLog converts ddl event to redo log
 func (d *DDLEvent) ToRedoLog() *RedoLog {
+	columns := make([]*ColumnInfo, 0, len(d.TableInfo.GetColumns()))
+	for _, col := range d.TableInfo.GetColumns() {
+		columns = append(columns, &ColumnInfo{
+			Name:               col.Name.String(),
+			OriginDefaultValue: col.GetOriginDefaultValue(),
+			Type:               col.GetType(),
+			Version:            col.Version,
+		})
+	}
 	redoLog := &RedoLog{
 		RedoDDL: &RedoDDLEvent{
 			DDL: &DDLEventInRedoLog{
 				StartTs:           d.GetStartTs(),
 				CommitTs:          d.GetCommitTs(),
 				Query:             d.Query,
+				Columns:           columns,
 				BlockedTables:     d.BlockedTables,
 				BlockedTableNames: d.BlockedTableNames,
 				NeedDroppedTables: d.NeedDroppedTables,
@@ -365,10 +388,23 @@ func (r *RedoDDLEvent) ToDDLEvent() *DDLEvent {
 		blockedTables = &InfluencedTables{InfluenceType: InfluenceTypeNormal}
 		blockedTableNames = []SchemaTableName{{SchemaName: r.TableName.Schema, TableName: r.TableName.Table}}
 	}
+	columns := make([]*timodel.ColumnInfo, 0, len(r.DDL.Columns))
+	for _, col := range r.DDL.Columns {
+		colInfo := &timodel.ColumnInfo{
+			ID:      int64(len(columns)),
+			Name:    ast.NewCIStr(col.Name),
+			State:   timodel.StatePublic,
+			Version: col.Version,
+		}
+		colInfo.SetType(col.Type)
+		colInfo.SetOriginDefaultValue(col.OriginDefaultValue)
+		columns = append(columns, colInfo)
+	}
 	return &DDLEvent{
-		TableInfo: &commonType.TableInfo{
-			TableName: r.TableName,
-		},
+		TableInfo: commonType.NewTableInfo4Decoder(r.TableName.Schema, &timodel.TableInfo{
+			Name:    ast.NewCIStr(r.TableName.Table),
+			Columns: columns,
+		}),
 		Query:             r.DDL.Query,
 		Type:              r.Type,
 		SchemaName:        r.TableName.Schema,

--- a/pkg/common/event/redo.go
+++ b/pkg/common/event/redo.go
@@ -84,16 +84,12 @@ type DDLEventInRedoLog struct {
 	NeedAddedTables   []Table           `msg:"need_added_tables"`
 }
 
+// ColumnInfo is for column meta in DDL event
 type ColumnInfo struct {
 	Name               string `msg:"name"`
 	OriginDefaultValue any    `msg:"origin_default"`
 	Type               byte   `msg:"type"`
-	// Version means the version of the column info.
-	// Version = 0: For OriginDefaultValue and DefaultValue of timestamp column will stores the default time in system time zone.
-	//              That is a bug if multiple TiDB servers in different system time zone.
-	// Version = 1: For OriginDefaultValue and DefaultValue of timestamp column will stores the default time in UTC time zone.
-	//              This will fix bug in version 0. For compatibility with version 0, we add version field in column info struct.
-	Version uint64 `msg:"version"`
+	Version            uint64 `msg:"version"`
 }
 
 // RedoColumn is for column meta

--- a/pkg/common/event/redo.go
+++ b/pkg/common/event/redo.go
@@ -216,14 +216,17 @@ func (r *RedoRowEvent) ToRedoLog() *RedoLog {
 
 // ToRedoLog converts ddl event to redo log
 func (d *DDLEvent) ToRedoLog() *RedoLog {
-	columns := make([]*ColumnInfo, 0, len(d.TableInfo.GetColumns()))
-	for _, col := range d.TableInfo.GetColumns() {
-		columns = append(columns, &ColumnInfo{
-			Name:               col.Name.String(),
-			OriginDefaultValue: col.GetOriginDefaultValue(),
-			Type:               col.GetType(),
-			Version:            col.Version,
-		})
+	var columns []*ColumnInfo
+	if d.TableInfo != nil {
+		columns = make([]*ColumnInfo, 0, len(d.TableInfo.GetColumns()))
+		for _, col := range d.TableInfo.GetColumns() {
+			columns = append(columns, &ColumnInfo{
+				Name:               col.Name.String(),
+				OriginDefaultValue: col.GetOriginDefaultValue(),
+				Type:               col.GetType(),
+				Version:            col.Version,
+			})
+		}
 	}
 	redoLog := &RedoLog{
 		RedoDDL: &RedoDDLEvent{
@@ -397,14 +400,22 @@ func (r *RedoDDLEvent) ToDDLEvent() *DDLEvent {
 			Version: col.Version,
 		}
 		colInfo.SetType(col.Type)
-		colInfo.SetOriginDefaultValue(col.OriginDefaultValue)
+		if err := colInfo.SetOriginDefaultValue(col.OriginDefaultValue); err != nil {
+			log.Panic("set origin default value failed",
+				zap.String("column", col.Name),
+				zap.Any("originDefaultValue", col.OriginDefaultValue),
+				zap.Error(err))
+		}
 		columns = append(columns, colInfo)
 	}
+	tableInfo := commonType.NewTableInfo4Decoder(r.TableName.Schema, &timodel.TableInfo{
+		ID:      r.TableName.TableID,
+		Name:    ast.NewCIStr(r.TableName.Table),
+		Columns: columns,
+	})
+	tableInfo.TableName.IsPartition = r.TableName.IsPartition
 	return &DDLEvent{
-		TableInfo: commonType.NewTableInfo4Decoder(r.TableName.Schema, &timodel.TableInfo{
-			Name:    ast.NewCIStr(r.TableName.Table),
-			Columns: columns,
-		}),
+		TableInfo:         tableInfo,
 		Query:             r.DDL.Query,
 		Type:              r.Type,
 		SchemaName:        r.TableName.Schema,

--- a/pkg/common/event/redo.go
+++ b/pkg/common/event/redo.go
@@ -390,7 +390,6 @@ func (r *RedoDDLEvent) ToDDLEvent() *DDLEvent {
 	columns := make([]*timodel.ColumnInfo, 0, len(r.DDL.Columns))
 	for _, col := range r.DDL.Columns {
 		colInfo := &timodel.ColumnInfo{
-			ID:      int64(len(columns)),
 			Name:    ast.NewCIStr(col.Name),
 			State:   timodel.StatePublic,
 			Version: col.Version,

--- a/pkg/common/event/redo_gen.go
+++ b/pkg/common/event/redo_gen.go
@@ -8,6 +8,188 @@ import (
 )
 
 // DecodeMsg implements msgp.Decodable
+func (z *ColumnInfo) DecodeMsg(dc *msgp.Reader) (err error) {
+	var field []byte
+	_ = field
+	var zb0001 uint32
+	zb0001, err = dc.ReadMapHeader()
+	if err != nil {
+		err = msgp.WrapError(err)
+		return
+	}
+	for zb0001 > 0 {
+		zb0001--
+		field, err = dc.ReadMapKeyPtr()
+		if err != nil {
+			err = msgp.WrapError(err)
+			return
+		}
+		switch msgp.UnsafeString(field) {
+		case "name":
+			z.Name, err = dc.ReadString()
+			if err != nil {
+				err = msgp.WrapError(err, "Name")
+				return
+			}
+		case "origin_default":
+			z.OriginDefaultValue, err = dc.ReadIntf()
+			if err != nil {
+				err = msgp.WrapError(err, "OriginDefaultValue")
+				return
+			}
+		case "type":
+			z.Type, err = dc.ReadByte()
+			if err != nil {
+				err = msgp.WrapError(err, "Type")
+				return
+			}
+		case "version":
+			z.Version, err = dc.ReadUint64()
+			if err != nil {
+				err = msgp.WrapError(err, "Version")
+				return
+			}
+		default:
+			err = dc.Skip()
+			if err != nil {
+				err = msgp.WrapError(err)
+				return
+			}
+		}
+	}
+	return
+}
+
+// EncodeMsg implements msgp.Encodable
+func (z *ColumnInfo) EncodeMsg(en *msgp.Writer) (err error) {
+	// map header, size 4
+	// write "name"
+	err = en.Append(0x84, 0xa4, 0x6e, 0x61, 0x6d, 0x65)
+	if err != nil {
+		return
+	}
+	err = en.WriteString(z.Name)
+	if err != nil {
+		err = msgp.WrapError(err, "Name")
+		return
+	}
+	// write "origin_default"
+	err = en.Append(0xae, 0x6f, 0x72, 0x69, 0x67, 0x69, 0x6e, 0x5f, 0x64, 0x65, 0x66, 0x61, 0x75, 0x6c, 0x74)
+	if err != nil {
+		return
+	}
+	err = en.WriteIntf(z.OriginDefaultValue)
+	if err != nil {
+		err = msgp.WrapError(err, "OriginDefaultValue")
+		return
+	}
+	// write "type"
+	err = en.Append(0xa4, 0x74, 0x79, 0x70, 0x65)
+	if err != nil {
+		return
+	}
+	err = en.WriteByte(z.Type)
+	if err != nil {
+		err = msgp.WrapError(err, "Type")
+		return
+	}
+	// write "version"
+	err = en.Append(0xa7, 0x76, 0x65, 0x72, 0x73, 0x69, 0x6f, 0x6e)
+	if err != nil {
+		return
+	}
+	err = en.WriteUint64(z.Version)
+	if err != nil {
+		err = msgp.WrapError(err, "Version")
+		return
+	}
+	return
+}
+
+// MarshalMsg implements msgp.Marshaler
+func (z *ColumnInfo) MarshalMsg(b []byte) (o []byte, err error) {
+	o = msgp.Require(b, z.Msgsize())
+	// map header, size 4
+	// string "name"
+	o = append(o, 0x84, 0xa4, 0x6e, 0x61, 0x6d, 0x65)
+	o = msgp.AppendString(o, z.Name)
+	// string "origin_default"
+	o = append(o, 0xae, 0x6f, 0x72, 0x69, 0x67, 0x69, 0x6e, 0x5f, 0x64, 0x65, 0x66, 0x61, 0x75, 0x6c, 0x74)
+	o, err = msgp.AppendIntf(o, z.OriginDefaultValue)
+	if err != nil {
+		err = msgp.WrapError(err, "OriginDefaultValue")
+		return
+	}
+	// string "type"
+	o = append(o, 0xa4, 0x74, 0x79, 0x70, 0x65)
+	o = msgp.AppendByte(o, z.Type)
+	// string "version"
+	o = append(o, 0xa7, 0x76, 0x65, 0x72, 0x73, 0x69, 0x6f, 0x6e)
+	o = msgp.AppendUint64(o, z.Version)
+	return
+}
+
+// UnmarshalMsg implements msgp.Unmarshaler
+func (z *ColumnInfo) UnmarshalMsg(bts []byte) (o []byte, err error) {
+	var field []byte
+	_ = field
+	var zb0001 uint32
+	zb0001, bts, err = msgp.ReadMapHeaderBytes(bts)
+	if err != nil {
+		err = msgp.WrapError(err)
+		return
+	}
+	for zb0001 > 0 {
+		zb0001--
+		field, bts, err = msgp.ReadMapKeyZC(bts)
+		if err != nil {
+			err = msgp.WrapError(err)
+			return
+		}
+		switch msgp.UnsafeString(field) {
+		case "name":
+			z.Name, bts, err = msgp.ReadStringBytes(bts)
+			if err != nil {
+				err = msgp.WrapError(err, "Name")
+				return
+			}
+		case "origin_default":
+			z.OriginDefaultValue, bts, err = msgp.ReadIntfBytes(bts)
+			if err != nil {
+				err = msgp.WrapError(err, "OriginDefaultValue")
+				return
+			}
+		case "type":
+			z.Type, bts, err = msgp.ReadByteBytes(bts)
+			if err != nil {
+				err = msgp.WrapError(err, "Type")
+				return
+			}
+		case "version":
+			z.Version, bts, err = msgp.ReadUint64Bytes(bts)
+			if err != nil {
+				err = msgp.WrapError(err, "Version")
+				return
+			}
+		default:
+			bts, err = msgp.Skip(bts)
+			if err != nil {
+				err = msgp.WrapError(err)
+				return
+			}
+		}
+	}
+	o = bts
+	return
+}
+
+// Msgsize returns an upper bound estimate of the number of bytes occupied by the serialized message
+func (z *ColumnInfo) Msgsize() (s int) {
+	s = 1 + 5 + msgp.StringPrefixSize + len(z.Name) + 15 + msgp.GuessSize(z.OriginDefaultValue) + 5 + msgp.ByteSize + 8 + msgp.Uint64Size
+	return
+}
+
+// DecodeMsg implements msgp.Decodable
 func (z *DDLEventInRedoLog) DecodeMsg(dc *msgp.Reader) (err error) {
 	var field []byte
 	_ = field
@@ -43,6 +225,37 @@ func (z *DDLEventInRedoLog) DecodeMsg(dc *msgp.Reader) (err error) {
 				err = msgp.WrapError(err, "Query")
 				return
 			}
+		case "columns":
+			var zb0002 uint32
+			zb0002, err = dc.ReadArrayHeader()
+			if err != nil {
+				err = msgp.WrapError(err, "Columns")
+				return
+			}
+			if cap(z.Columns) >= int(zb0002) {
+				z.Columns = (z.Columns)[:zb0002]
+			} else {
+				z.Columns = make([]*ColumnInfo, zb0002)
+			}
+			for za0001 := range z.Columns {
+				if dc.IsNil() {
+					err = dc.ReadNil()
+					if err != nil {
+						err = msgp.WrapError(err, "Columns", za0001)
+						return
+					}
+					z.Columns[za0001] = nil
+				} else {
+					if z.Columns[za0001] == nil {
+						z.Columns[za0001] = new(ColumnInfo)
+					}
+					err = z.Columns[za0001].DecodeMsg(dc)
+					if err != nil {
+						err = msgp.WrapError(err, "Columns", za0001)
+						return
+					}
+				}
+			}
 		case "blocked-tables":
 			if dc.IsNil() {
 				err = dc.ReadNil()
@@ -62,21 +275,21 @@ func (z *DDLEventInRedoLog) DecodeMsg(dc *msgp.Reader) (err error) {
 				}
 			}
 		case "blocked-table-names":
-			var zb0002 uint32
-			zb0002, err = dc.ReadArrayHeader()
+			var zb0003 uint32
+			zb0003, err = dc.ReadArrayHeader()
 			if err != nil {
 				err = msgp.WrapError(err, "BlockedTableNames")
 				return
 			}
-			if cap(z.BlockedTableNames) >= int(zb0002) {
-				z.BlockedTableNames = (z.BlockedTableNames)[:zb0002]
+			if cap(z.BlockedTableNames) >= int(zb0003) {
+				z.BlockedTableNames = (z.BlockedTableNames)[:zb0003]
 			} else {
-				z.BlockedTableNames = make([]SchemaTableName, zb0002)
+				z.BlockedTableNames = make([]SchemaTableName, zb0003)
 			}
-			for za0001 := range z.BlockedTableNames {
-				err = z.BlockedTableNames[za0001].DecodeMsg(dc)
+			for za0002 := range z.BlockedTableNames {
+				err = z.BlockedTableNames[za0002].DecodeMsg(dc)
 				if err != nil {
-					err = msgp.WrapError(err, "BlockedTableNames", za0001)
+					err = msgp.WrapError(err, "BlockedTableNames", za0002)
 					return
 				}
 			}
@@ -99,21 +312,21 @@ func (z *DDLEventInRedoLog) DecodeMsg(dc *msgp.Reader) (err error) {
 				}
 			}
 		case "need_added_tables":
-			var zb0003 uint32
-			zb0003, err = dc.ReadArrayHeader()
+			var zb0004 uint32
+			zb0004, err = dc.ReadArrayHeader()
 			if err != nil {
 				err = msgp.WrapError(err, "NeedAddedTables")
 				return
 			}
-			if cap(z.NeedAddedTables) >= int(zb0003) {
-				z.NeedAddedTables = (z.NeedAddedTables)[:zb0003]
+			if cap(z.NeedAddedTables) >= int(zb0004) {
+				z.NeedAddedTables = (z.NeedAddedTables)[:zb0004]
 			} else {
-				z.NeedAddedTables = make([]Table, zb0003)
+				z.NeedAddedTables = make([]Table, zb0004)
 			}
-			for za0002 := range z.NeedAddedTables {
-				err = z.NeedAddedTables[za0002].DecodeMsg(dc)
+			for za0003 := range z.NeedAddedTables {
+				err = z.NeedAddedTables[za0003].DecodeMsg(dc)
 				if err != nil {
-					err = msgp.WrapError(err, "NeedAddedTables", za0002)
+					err = msgp.WrapError(err, "NeedAddedTables", za0003)
 					return
 				}
 			}
@@ -130,9 +343,9 @@ func (z *DDLEventInRedoLog) DecodeMsg(dc *msgp.Reader) (err error) {
 
 // EncodeMsg implements msgp.Encodable
 func (z *DDLEventInRedoLog) EncodeMsg(en *msgp.Writer) (err error) {
-	// map header, size 7
+	// map header, size 8
 	// write "start-ts"
-	err = en.Append(0x87, 0xa8, 0x73, 0x74, 0x61, 0x72, 0x74, 0x2d, 0x74, 0x73)
+	err = en.Append(0x88, 0xa8, 0x73, 0x74, 0x61, 0x72, 0x74, 0x2d, 0x74, 0x73)
 	if err != nil {
 		return
 	}
@@ -161,6 +374,30 @@ func (z *DDLEventInRedoLog) EncodeMsg(en *msgp.Writer) (err error) {
 		err = msgp.WrapError(err, "Query")
 		return
 	}
+	// write "columns"
+	err = en.Append(0xa7, 0x63, 0x6f, 0x6c, 0x75, 0x6d, 0x6e, 0x73)
+	if err != nil {
+		return
+	}
+	err = en.WriteArrayHeader(uint32(len(z.Columns)))
+	if err != nil {
+		err = msgp.WrapError(err, "Columns")
+		return
+	}
+	for za0001 := range z.Columns {
+		if z.Columns[za0001] == nil {
+			err = en.WriteNil()
+			if err != nil {
+				return
+			}
+		} else {
+			err = z.Columns[za0001].EncodeMsg(en)
+			if err != nil {
+				err = msgp.WrapError(err, "Columns", za0001)
+				return
+			}
+		}
+	}
 	// write "blocked-tables"
 	err = en.Append(0xae, 0x62, 0x6c, 0x6f, 0x63, 0x6b, 0x65, 0x64, 0x2d, 0x74, 0x61, 0x62, 0x6c, 0x65, 0x73)
 	if err != nil {
@@ -188,10 +425,10 @@ func (z *DDLEventInRedoLog) EncodeMsg(en *msgp.Writer) (err error) {
 		err = msgp.WrapError(err, "BlockedTableNames")
 		return
 	}
-	for za0001 := range z.BlockedTableNames {
-		err = z.BlockedTableNames[za0001].EncodeMsg(en)
+	for za0002 := range z.BlockedTableNames {
+		err = z.BlockedTableNames[za0002].EncodeMsg(en)
 		if err != nil {
-			err = msgp.WrapError(err, "BlockedTableNames", za0001)
+			err = msgp.WrapError(err, "BlockedTableNames", za0002)
 			return
 		}
 	}
@@ -222,10 +459,10 @@ func (z *DDLEventInRedoLog) EncodeMsg(en *msgp.Writer) (err error) {
 		err = msgp.WrapError(err, "NeedAddedTables")
 		return
 	}
-	for za0002 := range z.NeedAddedTables {
-		err = z.NeedAddedTables[za0002].EncodeMsg(en)
+	for za0003 := range z.NeedAddedTables {
+		err = z.NeedAddedTables[za0003].EncodeMsg(en)
 		if err != nil {
-			err = msgp.WrapError(err, "NeedAddedTables", za0002)
+			err = msgp.WrapError(err, "NeedAddedTables", za0003)
 			return
 		}
 	}
@@ -235,9 +472,9 @@ func (z *DDLEventInRedoLog) EncodeMsg(en *msgp.Writer) (err error) {
 // MarshalMsg implements msgp.Marshaler
 func (z *DDLEventInRedoLog) MarshalMsg(b []byte) (o []byte, err error) {
 	o = msgp.Require(b, z.Msgsize())
-	// map header, size 7
+	// map header, size 8
 	// string "start-ts"
-	o = append(o, 0x87, 0xa8, 0x73, 0x74, 0x61, 0x72, 0x74, 0x2d, 0x74, 0x73)
+	o = append(o, 0x88, 0xa8, 0x73, 0x74, 0x61, 0x72, 0x74, 0x2d, 0x74, 0x73)
 	o = msgp.AppendUint64(o, z.StartTs)
 	// string "commit-ts"
 	o = append(o, 0xa9, 0x63, 0x6f, 0x6d, 0x6d, 0x69, 0x74, 0x2d, 0x74, 0x73)
@@ -245,6 +482,20 @@ func (z *DDLEventInRedoLog) MarshalMsg(b []byte) (o []byte, err error) {
 	// string "query"
 	o = append(o, 0xa5, 0x71, 0x75, 0x65, 0x72, 0x79)
 	o = msgp.AppendString(o, z.Query)
+	// string "columns"
+	o = append(o, 0xa7, 0x63, 0x6f, 0x6c, 0x75, 0x6d, 0x6e, 0x73)
+	o = msgp.AppendArrayHeader(o, uint32(len(z.Columns)))
+	for za0001 := range z.Columns {
+		if z.Columns[za0001] == nil {
+			o = msgp.AppendNil(o)
+		} else {
+			o, err = z.Columns[za0001].MarshalMsg(o)
+			if err != nil {
+				err = msgp.WrapError(err, "Columns", za0001)
+				return
+			}
+		}
+	}
 	// string "blocked-tables"
 	o = append(o, 0xae, 0x62, 0x6c, 0x6f, 0x63, 0x6b, 0x65, 0x64, 0x2d, 0x74, 0x61, 0x62, 0x6c, 0x65, 0x73)
 	if z.BlockedTables == nil {
@@ -259,10 +510,10 @@ func (z *DDLEventInRedoLog) MarshalMsg(b []byte) (o []byte, err error) {
 	// string "blocked-table-names"
 	o = append(o, 0xb3, 0x62, 0x6c, 0x6f, 0x63, 0x6b, 0x65, 0x64, 0x2d, 0x74, 0x61, 0x62, 0x6c, 0x65, 0x2d, 0x6e, 0x61, 0x6d, 0x65, 0x73)
 	o = msgp.AppendArrayHeader(o, uint32(len(z.BlockedTableNames)))
-	for za0001 := range z.BlockedTableNames {
-		o, err = z.BlockedTableNames[za0001].MarshalMsg(o)
+	for za0002 := range z.BlockedTableNames {
+		o, err = z.BlockedTableNames[za0002].MarshalMsg(o)
 		if err != nil {
-			err = msgp.WrapError(err, "BlockedTableNames", za0001)
+			err = msgp.WrapError(err, "BlockedTableNames", za0002)
 			return
 		}
 	}
@@ -280,10 +531,10 @@ func (z *DDLEventInRedoLog) MarshalMsg(b []byte) (o []byte, err error) {
 	// string "need_added_tables"
 	o = append(o, 0xb1, 0x6e, 0x65, 0x65, 0x64, 0x5f, 0x61, 0x64, 0x64, 0x65, 0x64, 0x5f, 0x74, 0x61, 0x62, 0x6c, 0x65, 0x73)
 	o = msgp.AppendArrayHeader(o, uint32(len(z.NeedAddedTables)))
-	for za0002 := range z.NeedAddedTables {
-		o, err = z.NeedAddedTables[za0002].MarshalMsg(o)
+	for za0003 := range z.NeedAddedTables {
+		o, err = z.NeedAddedTables[za0003].MarshalMsg(o)
 		if err != nil {
-			err = msgp.WrapError(err, "NeedAddedTables", za0002)
+			err = msgp.WrapError(err, "NeedAddedTables", za0003)
 			return
 		}
 	}
@@ -326,6 +577,36 @@ func (z *DDLEventInRedoLog) UnmarshalMsg(bts []byte) (o []byte, err error) {
 				err = msgp.WrapError(err, "Query")
 				return
 			}
+		case "columns":
+			var zb0002 uint32
+			zb0002, bts, err = msgp.ReadArrayHeaderBytes(bts)
+			if err != nil {
+				err = msgp.WrapError(err, "Columns")
+				return
+			}
+			if cap(z.Columns) >= int(zb0002) {
+				z.Columns = (z.Columns)[:zb0002]
+			} else {
+				z.Columns = make([]*ColumnInfo, zb0002)
+			}
+			for za0001 := range z.Columns {
+				if msgp.IsNil(bts) {
+					bts, err = msgp.ReadNilBytes(bts)
+					if err != nil {
+						return
+					}
+					z.Columns[za0001] = nil
+				} else {
+					if z.Columns[za0001] == nil {
+						z.Columns[za0001] = new(ColumnInfo)
+					}
+					bts, err = z.Columns[za0001].UnmarshalMsg(bts)
+					if err != nil {
+						err = msgp.WrapError(err, "Columns", za0001)
+						return
+					}
+				}
+			}
 		case "blocked-tables":
 			if msgp.IsNil(bts) {
 				bts, err = msgp.ReadNilBytes(bts)
@@ -344,21 +625,21 @@ func (z *DDLEventInRedoLog) UnmarshalMsg(bts []byte) (o []byte, err error) {
 				}
 			}
 		case "blocked-table-names":
-			var zb0002 uint32
-			zb0002, bts, err = msgp.ReadArrayHeaderBytes(bts)
+			var zb0003 uint32
+			zb0003, bts, err = msgp.ReadArrayHeaderBytes(bts)
 			if err != nil {
 				err = msgp.WrapError(err, "BlockedTableNames")
 				return
 			}
-			if cap(z.BlockedTableNames) >= int(zb0002) {
-				z.BlockedTableNames = (z.BlockedTableNames)[:zb0002]
+			if cap(z.BlockedTableNames) >= int(zb0003) {
+				z.BlockedTableNames = (z.BlockedTableNames)[:zb0003]
 			} else {
-				z.BlockedTableNames = make([]SchemaTableName, zb0002)
+				z.BlockedTableNames = make([]SchemaTableName, zb0003)
 			}
-			for za0001 := range z.BlockedTableNames {
-				bts, err = z.BlockedTableNames[za0001].UnmarshalMsg(bts)
+			for za0002 := range z.BlockedTableNames {
+				bts, err = z.BlockedTableNames[za0002].UnmarshalMsg(bts)
 				if err != nil {
-					err = msgp.WrapError(err, "BlockedTableNames", za0001)
+					err = msgp.WrapError(err, "BlockedTableNames", za0002)
 					return
 				}
 			}
@@ -380,21 +661,21 @@ func (z *DDLEventInRedoLog) UnmarshalMsg(bts []byte) (o []byte, err error) {
 				}
 			}
 		case "need_added_tables":
-			var zb0003 uint32
-			zb0003, bts, err = msgp.ReadArrayHeaderBytes(bts)
+			var zb0004 uint32
+			zb0004, bts, err = msgp.ReadArrayHeaderBytes(bts)
 			if err != nil {
 				err = msgp.WrapError(err, "NeedAddedTables")
 				return
 			}
-			if cap(z.NeedAddedTables) >= int(zb0003) {
-				z.NeedAddedTables = (z.NeedAddedTables)[:zb0003]
+			if cap(z.NeedAddedTables) >= int(zb0004) {
+				z.NeedAddedTables = (z.NeedAddedTables)[:zb0004]
 			} else {
-				z.NeedAddedTables = make([]Table, zb0003)
+				z.NeedAddedTables = make([]Table, zb0004)
 			}
-			for za0002 := range z.NeedAddedTables {
-				bts, err = z.NeedAddedTables[za0002].UnmarshalMsg(bts)
+			for za0003 := range z.NeedAddedTables {
+				bts, err = z.NeedAddedTables[za0003].UnmarshalMsg(bts)
 				if err != nil {
-					err = msgp.WrapError(err, "NeedAddedTables", za0002)
+					err = msgp.WrapError(err, "NeedAddedTables", za0003)
 					return
 				}
 			}
@@ -412,15 +693,23 @@ func (z *DDLEventInRedoLog) UnmarshalMsg(bts []byte) (o []byte, err error) {
 
 // Msgsize returns an upper bound estimate of the number of bytes occupied by the serialized message
 func (z *DDLEventInRedoLog) Msgsize() (s int) {
-	s = 1 + 9 + msgp.Uint64Size + 10 + msgp.Uint64Size + 6 + msgp.StringPrefixSize + len(z.Query) + 15
+	s = 1 + 9 + msgp.Uint64Size + 10 + msgp.Uint64Size + 6 + msgp.StringPrefixSize + len(z.Query) + 8 + msgp.ArrayHeaderSize
+	for za0001 := range z.Columns {
+		if z.Columns[za0001] == nil {
+			s += msgp.NilSize
+		} else {
+			s += z.Columns[za0001].Msgsize()
+		}
+	}
+	s += 15
 	if z.BlockedTables == nil {
 		s += msgp.NilSize
 	} else {
 		s += z.BlockedTables.Msgsize()
 	}
 	s += 20 + msgp.ArrayHeaderSize
-	for za0001 := range z.BlockedTableNames {
-		s += z.BlockedTableNames[za0001].Msgsize()
+	for za0002 := range z.BlockedTableNames {
+		s += z.BlockedTableNames[za0002].Msgsize()
 	}
 	s += 20
 	if z.NeedDroppedTables == nil {
@@ -429,8 +718,8 @@ func (z *DDLEventInRedoLog) Msgsize() (s int) {
 		s += z.NeedDroppedTables.Msgsize()
 	}
 	s += 18 + msgp.ArrayHeaderSize
-	for za0002 := range z.NeedAddedTables {
-		s += z.NeedAddedTables[za0002].Msgsize()
+	for za0003 := range z.NeedAddedTables {
+		s += z.NeedAddedTables[za0003].Msgsize()
 	}
 	return
 }

--- a/pkg/common/event/redo_gen_test.go
+++ b/pkg/common/event/redo_gen_test.go
@@ -9,6 +9,119 @@ import (
 	"github.com/tinylib/msgp/msgp"
 )
 
+func TestMarshalUnmarshalColumnInfo(t *testing.T) {
+	v := ColumnInfo{}
+	bts, err := v.MarshalMsg(nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	left, err := v.UnmarshalMsg(bts)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(left) > 0 {
+		t.Errorf("%d bytes left over after UnmarshalMsg(): %q", len(left), left)
+	}
+
+	left, err = msgp.Skip(bts)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(left) > 0 {
+		t.Errorf("%d bytes left over after Skip(): %q", len(left), left)
+	}
+}
+
+func BenchmarkMarshalMsgColumnInfo(b *testing.B) {
+	v := ColumnInfo{}
+	b.ReportAllocs()
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		v.MarshalMsg(nil)
+	}
+}
+
+func BenchmarkAppendMsgColumnInfo(b *testing.B) {
+	v := ColumnInfo{}
+	bts := make([]byte, 0, v.Msgsize())
+	bts, _ = v.MarshalMsg(bts[0:0])
+	b.SetBytes(int64(len(bts)))
+	b.ReportAllocs()
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		bts, _ = v.MarshalMsg(bts[0:0])
+	}
+}
+
+func BenchmarkUnmarshalColumnInfo(b *testing.B) {
+	v := ColumnInfo{}
+	bts, _ := v.MarshalMsg(nil)
+	b.ReportAllocs()
+	b.SetBytes(int64(len(bts)))
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		_, err := v.UnmarshalMsg(bts)
+		if err != nil {
+			b.Fatal(err)
+		}
+	}
+}
+
+func TestEncodeDecodeColumnInfo(t *testing.T) {
+	v := ColumnInfo{}
+	var buf bytes.Buffer
+	msgp.Encode(&buf, &v)
+
+	m := v.Msgsize()
+	if buf.Len() > m {
+		t.Log("WARNING: TestEncodeDecodeColumnInfo Msgsize() is inaccurate")
+	}
+
+	vn := ColumnInfo{}
+	err := msgp.Decode(&buf, &vn)
+	if err != nil {
+		t.Error(err)
+	}
+
+	buf.Reset()
+	msgp.Encode(&buf, &v)
+	err = msgp.NewReader(&buf).Skip()
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+func BenchmarkEncodeColumnInfo(b *testing.B) {
+	v := ColumnInfo{}
+	var buf bytes.Buffer
+	msgp.Encode(&buf, &v)
+	b.SetBytes(int64(buf.Len()))
+	en := msgp.NewWriter(msgp.Nowhere)
+	b.ReportAllocs()
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		v.EncodeMsg(en)
+	}
+	en.Flush()
+}
+
+func BenchmarkDecodeColumnInfo(b *testing.B) {
+	v := ColumnInfo{}
+	var buf bytes.Buffer
+	msgp.Encode(&buf, &v)
+	b.SetBytes(int64(buf.Len()))
+	rd := msgp.NewEndlessReader(buf.Bytes(), b)
+	dc := msgp.NewReader(rd)
+	b.ReportAllocs()
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		err := v.DecodeMsg(dc)
+		if err != nil {
+			b.Fatal(err)
+		}
+	}
+}
+
 func TestMarshalUnmarshalDDLEventInRedoLog(t *testing.T) {
 	v := DDLEventInRedoLog{}
 	bts, err := v.MarshalMsg(nil)

--- a/pkg/common/event/redo_test.go
+++ b/pkg/common/event/redo_test.go
@@ -1,0 +1,115 @@
+// Copyright 2026 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package event
+
+import (
+	"testing"
+
+	commonType "github.com/pingcap/ticdc/pkg/common"
+	timodel "github.com/pingcap/tidb/pkg/meta/model"
+	"github.com/pingcap/tidb/pkg/parser/ast"
+	"github.com/pingcap/tidb/pkg/parser/mysql"
+	"github.com/pingcap/tidb/pkg/types"
+	"github.com/stretchr/testify/require"
+)
+
+func TestRedoDDLEventRoundTripPreservesColumnMetadata(t *testing.T) {
+	originalTableInfo := commonType.NewTableInfo4Decoder("test", &timodel.TableInfo{
+		ID:   1001,
+		Name: ast.NewCIStr("t_redo"),
+		Columns: []*timodel.ColumnInfo{
+			newRedoDDLTestColumn(t, 1, "id", mysql.TypeLonglong, nil, timodel.CurrLatestColumnInfoVersion),
+			newRedoDDLTestColumn(t, 2, "status", mysql.TypeVarchar, "ready", timodel.CurrLatestColumnInfoVersion),
+			newRedoDDLTestColumn(t, 3, "created_at", mysql.TypeTimestamp, "2024-01-02 03:04:05", timodel.ColumnInfoVersion0),
+			newRedoDDLTestColumn(t, 4, "flags", mysql.TypeBit, "1", timodel.CurrLatestColumnInfoVersion),
+		},
+	})
+	originalTableInfo.TableName.IsPartition = true
+
+	ddlEvent := &DDLEvent{
+		Type:              byte(timodel.ActionCreateTable),
+		Query:             "create table test.t_redo(id bigint, status varchar(16) default 'ready')",
+		TableInfo:         originalTableInfo,
+		StartTs:           11,
+		FinishedTs:        22,
+		BlockedTableNames: []SchemaTableName{{SchemaName: "test", TableName: "t_redo"}},
+	}
+
+	redoLog := ddlEvent.ToRedoLog()
+	data, err := redoLog.MarshalMsg(nil)
+	require.NoError(t, err)
+
+	var decoded RedoLog
+	left, err := decoded.UnmarshalMsg(data)
+	require.NoError(t, err)
+	require.Empty(t, left)
+
+	roundTrip := decoded.RedoDDL.ToDDLEvent()
+	require.Equal(t, ddlEvent.Query, roundTrip.Query)
+	require.Equal(t, ddlEvent.StartTs, roundTrip.StartTs)
+	require.Equal(t, ddlEvent.FinishedTs, roundTrip.FinishedTs)
+	require.Equal(t, ddlEvent.BlockedTableNames, roundTrip.BlockedTableNames)
+	require.Equal(t, originalTableInfo.TableName, roundTrip.TableInfo.TableName)
+
+	originalColumns := originalTableInfo.GetColumns()
+	roundTripColumns := roundTrip.TableInfo.GetColumns()
+	require.Len(t, roundTripColumns, len(originalColumns))
+	for i, originalColumn := range originalColumns {
+		roundTripColumn := roundTripColumns[i]
+		require.Equal(t, originalColumn.Name.O, roundTripColumn.Name.O)
+		require.Equal(t, originalColumn.GetType(), roundTripColumn.GetType())
+		require.Equal(t, originalColumn.GetOriginDefaultValue(), roundTripColumn.GetOriginDefaultValue())
+		require.Equal(t, originalColumn.Version, roundTripColumn.Version)
+	}
+}
+
+func TestDDLEventToRedoLogHandlesNilTableInfo(t *testing.T) {
+	ddlEvent := &DDLEvent{
+		Type:       byte(timodel.ActionCreateSchema),
+		Query:      "create database test_redo",
+		StartTs:    33,
+		FinishedTs: 44,
+	}
+
+	require.NotPanics(t, func() {
+		redoLog := ddlEvent.ToRedoLog()
+		require.Equal(t, ddlEvent.Query, redoLog.RedoDDL.DDL.Query)
+		require.Equal(t, ddlEvent.StartTs, redoLog.RedoDDL.DDL.StartTs)
+		require.Equal(t, ddlEvent.FinishedTs, redoLog.RedoDDL.DDL.CommitTs)
+		require.Nil(t, redoLog.RedoDDL.DDL.Columns)
+	})
+}
+
+func newRedoDDLTestColumn(
+	t *testing.T,
+	id int64,
+	name string,
+	tp byte,
+	originDefaultValue any,
+	version uint64,
+) *timodel.ColumnInfo {
+	t.Helper()
+
+	fieldType := types.NewFieldType(tp)
+	column := &timodel.ColumnInfo{
+		ID:        id,
+		Offset:    int(id - 1),
+		Name:      ast.NewCIStr(name),
+		State:     timodel.StatePublic,
+		FieldType: *fieldType,
+		Version:   version,
+	}
+	require.NoError(t, column.SetOriginDefaultValue(originDefaultValue))
+	return column
+}

--- a/pkg/common/event/redo_test.go
+++ b/pkg/common/event/redo_test.go
@@ -91,6 +91,22 @@ func TestDDLEventToRedoLogHandlesNilTableInfo(t *testing.T) {
 	})
 }
 
+func TestDDLEventToRedoLogHandlesUninitializedTableInfo(t *testing.T) {
+	ddlEvent := &DDLEvent{
+		Type:       byte(timodel.ActionAddColumn),
+		Query:      "alter table test_redo.t add column c int",
+		StartTs:    55,
+		FinishedTs: 66,
+		TableInfo:  &commonType.TableInfo{},
+	}
+
+	redoLog := ddlEvent.ToRedoLog()
+	require.Equal(t, ddlEvent.Query, redoLog.RedoDDL.DDL.Query)
+	require.Equal(t, ddlEvent.StartTs, redoLog.RedoDDL.DDL.StartTs)
+	require.Equal(t, ddlEvent.FinishedTs, redoLog.RedoDDL.DDL.CommitTs)
+	require.Empty(t, redoLog.RedoDDL.DDL.Columns)
+}
+
 func newRedoDDLTestColumn(
 	t *testing.T,
 	id int64,

--- a/pkg/common/table_info.go
+++ b/pkg/common/table_info.go
@@ -203,6 +203,9 @@ func (ti *TableInfo) ShadowCopyColumnSchema() *columnSchema {
 }
 
 func (ti *TableInfo) GetColumns() []*model.ColumnInfo {
+	if ti == nil || ti.columnSchema == nil {
+		return nil
+	}
 	return ti.columnSchema.Columns
 }
 

--- a/pkg/eventservice/event_broker_test.go
+++ b/pkg/eventservice/event_broker_test.go
@@ -1394,7 +1394,7 @@ func TestProcessTableTriggerDispatcherRespectsSyncPointBoundary(t *testing.T) {
 
 	changefeed := broker.getOrSetChangefeedStatus(info.GetChangefeedID(), info.GetSyncPointInterval())
 	dispatcher := newDispatcherStat(info, 1, 1, nil, changefeed)
-	dispatcher.seq.Store(1)
+	dispatcher.setHandshaked()
 
 	dispatcherPtr := &atomic.Pointer[dispatcherStat]{}
 	dispatcherPtr.Store(dispatcher)
@@ -1438,7 +1438,7 @@ func TestProcessTableTriggerDispatcherNudgesCommitStageWhenNoForwardRange(t *tes
 
 	changefeed := broker.getOrSetChangefeedStatus(info.GetChangefeedID(), info.GetSyncPointInterval())
 	dispatcher := newDispatcherStat(info, 1, 1, nil, changefeed)
-	dispatcher.seq.Store(1)
+	dispatcher.setHandshaked()
 	changefeed.syncPointPreparingTs.Store(ts10)
 	changefeed.syncPointInFlightTs.Store(ts10)
 

--- a/tests/integration_tests/ddl_default_current_timestamp/conf/changefeed.toml
+++ b/tests/integration_tests/ddl_default_current_timestamp/conf/changefeed.toml
@@ -1,0 +1,3 @@
+[consistent]
+level = "eventual"
+storage = "file:///tmp/tidb_cdc_test/ddl_default_current_timestamp/redo"

--- a/tests/integration_tests/ddl_default_current_timestamp/run.sh
+++ b/tests/integration_tests/ddl_default_current_timestamp/run.sh
@@ -74,6 +74,47 @@ function run() {
 
 	export GO_FAILPOINTS=''
 	cleanup_process $CDC_BINARY
+
+	# Start a redo-backed changefeed and block downstream DDL/DML execution so the
+	# following DDL must be applied by `cdc redo apply`.
+	redo_start_ts=$(run_cdc_cli_tso_query ${UP_PD_HOST_1} ${UP_PD_PORT_1})
+	export GO_FAILPOINTS='github.com/pingcap/ticdc/pkg/sink/mysql/MySQLSinkHangLongTime=return(true);github.com/pingcap/ticdc/pkg/sink/mysql/MySQLSinkExecDDLDelay=return("3600")'
+	run_cdc_server --workdir $WORK_DIR --binary $CDC_BINARY --config "$CUR/conf/server.toml" --logsuffix redo
+	cdc_cli_changefeed create --start-ts=$redo_start_ts --sink-uri="$SINK_URI" --config="$CUR/conf/changefeed.toml" -c "test"
+
+	run_sql "SET time_zone = '${TIME_ZONE}'; SET @@timestamp = 1000000004.222222; ALTER TABLE ${DB_NAME}.t_fp ADD COLUMN redo_c DATETIME(6) DEFAULT CURRENT_TIMESTAMP(6);" ${UP_TIDB_HOST} ${UP_TIDB_PORT}
+
+	storage_path="file://$WORK_DIR/redo"
+	tmp_download_path=$WORK_DIR/cdc_data/redo/test
+	current_tso=$(run_cdc_cli_tso_query ${UP_PD_HOST_1} ${UP_PD_PORT_1})
+	ensure 100 check_redo_resolved_ts test $current_tso $storage_path $tmp_download_path/meta
+
+	export GO_FAILPOINTS=''
+	cleanup_process $CDC_BINARY
+
+	down_redo_col_count=$(mysql -uroot -h${DOWN_TIDB_HOST} -P${DOWN_TIDB_PORT} --default-character-set utf8mb4 -Nse "SELECT COUNT(*) FROM information_schema.columns WHERE table_schema='${DB_NAME}' AND table_name='t_fp' AND column_name='redo_c';")
+	if [ "$down_redo_col_count" != "0" ]; then
+		echo "downstream redo_c should not exist before redo apply"
+		exit 1
+	fi
+
+	cdc redo apply --log-level debug --tmp-dir="$tmp_download_path/apply" \
+		--storage="$storage_path" \
+		--sink-uri="$SINK_URI" >$WORK_DIR/cdc_redo.log
+
+	redo_expected_ts="2001-09-09 09:46:44.222222"
+	down_redo_c=""
+	for i in $(seq 1 30); do
+		down_redo_c=$(mysql -uroot -h${DOWN_TIDB_HOST} -P${DOWN_TIDB_PORT} --default-character-set utf8mb4 --init-command="set time_zone='${TIME_ZONE}'" -Nse "SELECT redo_c FROM ${DB_NAME}.t_fp WHERE id=1;" 2>/dev/null || true)
+		if [ "$down_redo_c" = "$redo_expected_ts" ]; then
+			break
+		fi
+		sleep 2
+	done
+	if [ "$down_redo_c" != "$redo_expected_ts" ]; then
+		echo "redo apply should set redo_c to ${redo_expected_ts}, got ${down_redo_c}"
+		exit 1
+	fi
 }
 
 if [ "$SINK_TYPE" != "mysql" ]; then

